clang/lib/CodeGen/CGClass.cpp

```diff
@@ -246,7 +246,7 @@ ApplyNonVirtualAndVirtualOffset(CodeGenFunction &CGF, Address addr,
 
   // Apply the base offset.
   llvm::Value *ptr = addr.getPointer();
-  ptr = CGF.Builder.CreateBitCast(ptr, CGF.Int8PtrTy);
+  ptr = CGF.Builder.CreatePointerBitCastOrAddrSpaceCast(ptr, CGF.Int8PtrTy);
```

Comment:
Shouldn't you cast a pointer to an i8* pointer of the *same address space*? Casting from generic (or actually any other than private) address space to private is illegal, unless you know the actual address space of a pointer.

Reply:
Could you please provide a link to SPIRV/OpenCL specification which tells that cast from generic to private is illegal?

Reply:
There is no spec that defines OpenCL/SYCL address spaces in LLVM IR (unless you consider the old SPIR spec). 

Generally, if you have a generic pointer and you want to cast it to global/local/private, you should do this if and only if you know for sure that this generic pointer is in fact global, local, or private, and it can never be in any other address space. In other words, you have to preserve the original address space of a pointer or leave it generic. Otherwise, you're basically lying to the compiler, and this could result in miscompilation, or an inference pass can go crazy when it finds out the original address space and a mismatch between source and dest address spaces in an addrspacecast instruction.

Reply:
So, seems that actually it's not illegal, but it probably could lead to address space inference failure.
I was a little bit confused, because OpenCL allows casts from generic address space to named address spaces and I think it's easy to get address space cast from generic address space to named address space in IR generated from OpenCL.
Okay, I will change this patch to emit int8* pointer of the same address space for safety reason.